kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java

```diff
@@ -202,6 +204,16 @@ public Transaction build(Engine engine) {
       metadata = metadataWithCMInfo.get();
     }
 
+    if (schema.isPresent() && !isNewTable) {
+      SchemaUtils.validateUpdatedSchema(
+          snapshot.getSchema(),
+          schema.get(),
+          snapshot.getPartitionColumnNames(),
+          snapshot.getMetadata());
```

Comment:
Does this need to use the latest metadata? what if I enable IcebergCompatV2 in this same txn?

If yes, please add a test case for this

Reply:
I'm working on a PR that reorganizes some of this to hopefully avoid confusion like this

Reply:
https://github.com/delta-io/delta/pull/4267

Reply:
Fixed and added a test, for now what I did was largely revert to the existing state of TransactionBuilderImpl with the exception of the updatedSchema case which is validated with the latest metadata. Once your PR gets in, I'll rebase and put those checks in the right place with the correct metadata.